kotlin-native/runtime/src/main/kotlin/kotlin/native/Annotations.kt

```diff
@@ -83,3 +83,67 @@ public annotation class EagerInitialization
 @Target(AnnotationTarget.FUNCTION)
 @Retention(AnnotationRetention.BINARY)
 public actual annotation class CName(actual val externName: String = "", actual val shortName: String = "")
+
+/**
+ * This annotation marks the experimental Objective-C export refinement annotations.
+ */
+@RequiresOptIn
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+public actual annotation class ExperimentalObjCRefinement
+
+/**
+ * Meta-annotation that instructs the Kotlin compiler to remove the annotated function or property from the public Objective-C API.
+ *
+ * Annotation processors that refine the public Objective-C API can annotate their annotations with this meta-annotation
+ * to have the original declarations automatically removed from the public API.
+ *
+ * Note: only annotations with [AnnotationTarget.FUNCTION] and/or [AnnotationTarget.PROPERTY] are supported.
+ */
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@ExperimentalObjCRefinement
+public actual annotation class RefinesForObjC
+
+/**
+ * Instructs the Kotlin compiler to remove this function or property from the public Objective-C API.
+ */
+@RefinesForObjC
+@Target(AnnotationTarget.PROPERTY, AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@ExperimentalObjCRefinement
+public actual annotation class RefinedForObjC
+
+/**
+ * Meta-annotation that instructs the Kotlin compiler to mark the annotated function or property as
+ * `swift_private` in the generated Objective-C API.
+ *
+ * Annotation processors that refine the public API in Swift can annotate their annotations with this meta-annotation
+ * to automatically hide the annotated declarations from Swift.
+ *
+ * See Apple's documentation of the [`NS_REFINED_FOR_SWIFT`](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/improving_objective-c_api_declarations_for_swift)
+ * macro for more information on refining Objective-C declarations in Swift.
+ *
+ * Note: only annotations with [AnnotationTarget.FUNCTION] and/or [AnnotationTarget.PROPERTY] are supported.
+ */
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@ExperimentalObjCRefinement
+public actual annotation class RefinesInSwift
+
+/**
+ * Instructs the Kotlin compiler to mark this function or property as `swift_private` in the generated Objective-C API.
+ *
+ * See Apple's documentation of the [`NS_REFINED_FOR_SWIFT`](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/improving_objective-c_api_declarations_for_swift)
+ * macro for more information on refining Objective-C declarations in Swift.
+ */
+@RefinesInSwift
+@Target(AnnotationTarget.PROPERTY, AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@ExperimentalObjCRefinement
+public actual annotation class RefinedInSwift
```

Comment:
`HidesFromObjC`/`HiddenFromObjC` seem a good option: it is correct and clear enough.

As for `RefinedInSwift`: unfortunately, there is no common term that properly describes the behavior. Swift uses "refined" in its documentation. Under the hood, it is defined as "private", like
```c
# define CF_REFINED_FOR_SWIFT __attribute__((swift_private))

#define NS_REFINED_FOR_SWIFT CF_REFINED_FOR_SWIFT
```

But, the [documentation](https://clang.llvm.org/docs/AttributeReference.html#swift-private) for `swift_private` seems wrong: it says that "Declarations marked with the swift_private attribute are hidden from the framework client", but the test in this PR shows the opposite: [framework](https://github.com/JetBrains/kotlin/blob/bce06e6aa1df38784339e3e553508954eb87f41c/kotlin-native/backend.native/tests/objcexport/refined.kt#L28), [client](https://github.com/JetBrains/kotlin/blob/bce06e6aa1df38784339e3e553508954eb87f41c/kotlin-native/backend.native/tests/objcexport/refined.swift#L5). So "private" doesn't accurately reflect the behaviour.

Finally, the term "private" is not used in the public [documentation](https://developer.apple.com/documentation/swift/improving-objective-c-api-declarations-for-swift) for `NS_REFINED_FOR_SWIFT`. So "private" is neither accurate nor commonly recognized.

I guess, this leaves us with "refined" as the best option. @ilya-g what do you think?

Reply:
Yes, it doesn't seem to be private, but the documentation states, at least as an intent, that it is "hidden" from the framework client. So "HiddenInSwift" could work. 
Although, as it turns, it is not hidden entirely, but just masked with `__` prefix, and we can consider some synonyms, like `ConcealedInSwift`, `DisguisedInSwift`, `MaskedInSwift`, I don't think the conveyed difference in meaning is enough to use less common words for that.

Reply:
One concern I have with naming both annotations "Hidden" is that the meaning of hidden isn't the same. For ObjC it's more like removed (from the header) than hidden.

But like I said it seems there isn't a "perfect" name.
So maybe we should just pick some names and use the documentation to further explain it's use/meaning/effect?

Personally I would stick to the "refined" name for the Swift annotations (to keep them aligned with the macro Apple uses).

For the ObjC annotations we could also consider something like `@ObjCInternal` (although I am not sure what the meta annotation name would be).

Reply:
> One concern I have with naming both annotations "Hidden" is that the meaning of hidden isn't the same.

I agree.

> For ObjC it's more like removed (from the header) than hidden.

@rickclephas What's the difference between "removed" and "hidden"? I mean, if we didn't have the second annotation, why wouldn't `HiddenFromObjC` work?

> like `ConcealedInSwift`, `DisguisedInSwift`, `MaskedInSwift`, I don't think the conveyed difference in meaning is enough to use less common words for that.

@ilya-g why is "refined" less common than "concealed", "masked" or "disguised"? I mean, at least "refined" is used in Swift.

Reply:
> @rickclephas What's the difference between "removed" and "hidden"? I mean, if we didn't have the second annotation, why wouldn't `HiddenFromObjC` work?

To me "removed" means something is completely gone.
On the other hand "hidden" would mean "I can't see it", but it still exists.

If we are talking about the public API then:
- the ObjC annotation removes declarations
- the Swift annotation "renames" declarations resulting in those declarations being hidden

Reply:
@SvyatoslavScherbina I meant that "concealed", "masked", "disguised" are less common than "hidden". 
Regarding "refined" I still think that it doesn't describe what actually happens.

Reply:
> To me "removed" means something is completely gone.
On the other hand "hidden" would mean "I can't see it", but it still exists.

@rickclephas Well, with `RefinedForObjC`, Kotlin declaration still exists, but Obj-C can't see it. So, why isn't it "hidden"?

> I meant that "concealed", "masked", "disguised" are less common than "hidden".
Regarding "refined" I still think that it doesn't describe what actually happens.

@ilya-g Ah, I see. So, your proposal is to use "hidden" for both Swift and Obj-C annotations, right?

Reply:
> @rickclephas Well, with RefinedForObjC, Kotlin declaration still exists, but Obj-C can't see it. So, why isn't it "hidden"?

Because you can't use that declaration from ObjC.
It's no longer part of the ObjC API. Where in case of the Swift annotation it still is.

But now that I am thinking about this some more.
The behavior of the ObjC annotation is similar to the `HIDDEN` deprecation level.

I just feel like the Swift annotation has a somewhat different behavior (and possibly different usecases).

Reply:
I understand that behaviour differs for Swift and Obj-C annotations, and this makes using "hidden" for both feel weird. That's not what I ask.
You said:
> On the other hand "hidden" would mean "I can't see it", but it still exists.

But that's exactly what `RefinedForObjC` does: you can't see this Kotlin declaration from Obj-C, but it still exists (in Kotlin).

> Because you can't use that declaration from ObjC.
> It's no longer part of the ObjC API. Where in case of the Swift annotation it still is.

"ObjC API" is the implementation detail. You can think of it like Kotlin declaration is directly available from Obj-C instead.

Reply:
Correct. I am not saying that `@HiddenFromObjC` would be a bad name. On the contrary, I actually think it makes perfect sense!

I am just saying that both Swift and ObjC could use the hidden name, but IMO they can't both be named hidden.

Reply:
Alright so I think we all agree on the the following names for the ObjC annotations, right?:
* `@HidesFromObjC`: meta-annotation that will hide declarations from ObjC by removing them from the header 
* `@HiddenFromObjC`: built-in `@HidesFromObjC` annotation that can by used by developers

Regarding the Swift annotations: it seems the major concern is with the implied "code generation".
Although the term "refine" might be less known/common than "hidden", it does align with Apples documentation.

So how about the following names for the Swift annotations?:
* `@RefinesInSwift`: meta-annotation that will mark ObjC declarations as `swift_private`, allowing a more Swift like version to be created/generated
* `@RefineInSwift`: built-in `@RefinesInSwift` annotation that can be used by developers which doesn't auto generated any code.

IMO it describes best what will happen. However I am not sure if the single `s` is enough of a difference.
It's not much of a difference, but they can't actually be used on the same type of declarations 🤔.

Reply:
Is it `RefineInSwift` or `RefinedInSwift`? If the former, why is it different from the original version?

Reply:
It's`RefineInSwift`.

The difference being that `Refined` is an adjective, so it basically means that the declaration is refined in Swift. Which could imply that some kind of code gen is present (which isn't the case).

On the other hand `Refine` is a verb, so the developer should provided/create the refinement.

Update about the minor differences in naming: that was also the case for `Refines` and `Refined`.

Reply:
Ah, ok, thank you for the explanation!
@ilya-g what do you think about `RefineInSwift`?

Reply:
Hmm, might not be that great after all.
`@Suppress` is also a verb, but in that case you tell the compiler to suppress a warning.

I guess it depends on how you look at it 😅.
In that case I think I would prefer to tell the compiler that I have refined the declaration with `@RefinedInSwift`.

P.S. sorry was hoping to make the decision easier, not harder...

Reply:
Ok. Then, at this point, I would prefer it to be `HidesFromSwift` and `HiddenFromSwift`.
In particular, for a user of your plugin, the difference in meaning of "hidden" between `HidesFromSwift` and `HidesFromObjC` would typically be non-observable -- both hide the declaration. Yes, with `HidesFromSwift` you still can access the declaration, but does this matter if the intention is to let the plugin generate the wrapper, not ask the user to write that wrapper manually?

Reply:
No, in that case that probably wouldn't matter. However I think the primary use case for the Swift annotation is to write a "wrapper". In case of `HiddenFromSwift` that would mean the primary use case would be for the user to write a wrapper. I can't think of a case where you would want to expose the ObjC API, but hide the Swift one (without writing a more Swift friendly version).

Depending on the use case all these names could be valid.
I guess the question we need to answer is what is the primary use case for each annotation.
Some thoughts on that:

`Refines`/`Hides` has 2 use cases:
1. annotating the build-in annotation `Refined`/`Hidden`
2. annotating annotations from code gen plugins (most cases will fall into this category)

`RefinedForObjC`/`HiddenFromObjC` has 2 use cases:
1. annotating declarations that the user wants to hide from ObjC
2. annotating declarations that the user "refined" for ObjC (in Kotlin)

`RefinedInSwift`/`HiddenFromSwift` has only a single use case:
1. annotating declarations that the user "refined" for/in Swift

That's why I think:
* `HiddenFromSwift` doesn't make sense
* `RefinedForObjC`/`HiddenFromObjC` could both make sense depending on what the primary use case is
* `Hides` makes less sense since most of the annotated annotations are actually refining the API (with the built-in annotation as exception)

Anyway that's just my opinion, I can live with any of these names 🙂.

Reply:
If we don't want to give up "Refine" for swift annotations, I can propose another naming idea. Since the intent of `RefinedInSwift` is that the user should provide a swift-friendly wrapper, it could be named `ShouldRefineInSwift` — wordy, but the intent is clear that the task to make an actual refinement is on the user.

On the other hand, `RefinesInSwift` name for the meta annotation is fine, because usually it would be applied on annotations for annotation processors which would provide that refinement.